Initial functionality: Endpoints for cleaning field names and merging table versions

jacobmpeters opened the pull request:
- modules:
    - `core.endpoints`
       -  `fix_loop_variables`: to fix field names for loop questions responses
       - `merge_table_versions`:  merge versions of baseline questionnaire tables using full outer join, coalescing redundant fields as needed
     - `core.constants`: stores string constants used elsewhere in the code base that should be configured in one place
     - `core.utils`: low level helper functions for interacting with bigquery, handling strings, etcetera
     - `core.query_composition`: main functions to build queries used in endpoints
     - `core.variable_normalizer`: functions to check/fix variable names
     - `core.request_helpers`: functions for validating data/params sent to endpoints
- ci/cd and testing:
   - bin: bash scripts for local testing of endpoints
   - Dockerfile
   - cloudbuild.yaml: configured to deploy Docker container and initialize Cloud Run
  
Closes https://github.com/Analyticsphere/pr2-documentation/issues/7
Closes https://github.com/Analyticsphere/pr2-documentation/issues/8

https://github.com/episphere/connect/issues/1242

jacobmpeters committed:
initial functionality

jacobmpeters committed:
add cloudbuild

eddy-ccc commented on core/__init__.py:
Copy the .gitignore from the file process or orchestrator into this project, it'll exclude these .DS_Store files. Preventing them from getting into the git history is 100x easier than removing them after the fact.

jacobmpeters commented on core/constants.py:
```diff
@@ -0,0 +1,130 @@
+# -----------------------------------------------------------------------------
+
+SERVICE_NAME = "pr2-transformation"
+OUTPUT_SQL_PATH = "gs://pr2-pipeline-artifacts/sql"  # Can be local path or GCS Bucket, e.g., "gs://my-bucket/queries"
```
break up into "artifacts" bucket and add the SQL folder using fstring.

jacobmpeters replied:
Resolved: [0400c42](https://github.com/Analyticsphere/pr2-transformation/pull/1/commits/0400c423935bdedc8f6f608889507eee8839c98f)

jacobmpeters commented on core/constants.py:
```diff
@@ -0,0 +1,130 @@
+        print(f"\nSource Table: {mapping['source']}")
+        print(f"Destination Table: {mapping['destination']}\n")
+        print("-" * 80)
+    
```
remove or separate into separate file

jacobmpeters replied:
resolved: [f49f76a](https://github.com/Analyticsphere/pr2-transformation/pull/1/commits/f49f76a17afa735f56387f9b1aea10980493db54)

jacobmpeters commented on core/endpoints.py:
```diff
@@ -0,0 +1,53 @@
+
+@app.route('/merge_table_versions', methods=['POST'])
+def merge_table_versions():
+    mapping: dict[str, any] = request.get_json() or {}
```
consider simplifying this, removing `any` and run mypi

jacobmpeters commented on core/query_composition.py:
```diff
@@ -0,0 +1,196 @@
+            raise ValueError(error_msg)
+        alias = f"v{idx}"
+        aliases.append(alias)
+        table_columns[alias] = set(cols)
```
Add comment to make sure that the order is preserved and that the order is important for Coalesce statements later on.

jacobmpeters replied:
This will be addressed in a separate issue: https://github.com/Analyticsphere/pr2-documentation/issues/11

jacobmpeters commented on core/query_composition.py:
```diff
@@ -0,0 +1,196 @@
+    from_clause = f"{base_table} {base_alias}"
+    
+    join_clauses = []
+    for alias, table in zip(aliases[:-1][::-1], source_tables[:-1][::-1]):
```
rewrite this in simpler syntax.

jacobmpeters replied:
make a third table to test with, even if it is a copy/mutated version of another table

jacobmpeters commented on core/utils.py:
```diff
@@ -0,0 +1,216 @@
+    table = client.get_table(fq_table)
+    return [schema_field.name for schema_field in table.schema]
+
+def save_sql_string(sql: str, path: str, storage_client: storage.Client = None) -> None:
```
put types on variables

jacobmpeters commented on core/utils.py:
```diff
@@ -0,0 +1,216 @@
+        False
+    """
+    allowed_var_names = ['Connect_ID', 'token', 'uid']
+    allowed_extras = ['SIBCANC3D','CHOL','MOMCANC3D','SIBCANC3O','UF','DADCANC3K','BLOODCLOT','DEPRESS2',
```
Put this in the constants file

jacobmpeters replied:
Resolved: [9694f61](https://github.com/Analyticsphere/pr2-transformation/pull/1/commits/9694f61944be16058458b190c3955d32193af023)

eddy-ccc commented on core/constants.py:
```diff
@@ -0,0 +1,130 @@
+# Each enum sets its dataset once as a class variable (_DATASET) so that
+# changes to the dataset require an update in only one place.
+
+class SourceTables(StrEnum):
```
Most of these constants should be values that get passed into the functionality within this repo via the orchestrator

jacobmpeters replied:
Good point. I'll address this in a separate issue.

eddy-ccc commented on core/utils.py:
```diff
@@ -0,0 +1,216 @@
+        # Remove the gs:// scheme and split bucket name from the rest of the path.
+        path_without_scheme = path[5:]
+        parts = path_without_scheme.split("/", 1)
+        if len(parts) != 2:
```
Will this function work if the file exists within multiple subdirectories? The example has gs://bucket_name/path/to/file, but based on the exception checking, it seems like it may only work with something like gs://bucket_name/file

jacobmpeters replied:
Thanks for checking. I double checked this and confirmed that this `path.split("/",1)` splits the string once by the first "/" so it will retrieve the bucket_name and everything that follows the first backslash, such as "path/to/file"

eddy-ccc commented on core/utils.py:
```diff
@@ -0,0 +1,216 @@
+        blob.upload_from_string(sql)
+        print(f"SQL saved to {path}")
+    else:
+        # Ensure the local directory exists.
```
I assume this is used only when testing the repo locally, just be careful you don't save anything to local directories when executing this as a Cloud Run function. Everything local to the VM executing the Cloud Run function will be destroyed when it's done running.

jacobmpeters replied:
Exactly. I just included that for local testing. Thanks for the warning.

eddy-ccc commented on core/utils.py:
```diff
@@ -0,0 +1,216 @@
+    """
+    if path.startswith("gs://"):
+        # Remove the gs:// scheme and split bucket name from the rest of the path.
+        path_without_scheme = path[5:]
```
What does the 5: do here?

jacobmpeters replied:
`path[5:]` gets all of the characters after the first 5, so everything after "gs://".

eddy-ccc replied:
Something like `path.replace('gs://', '')` may be a better option there, it's more clear and less fragile than removing a specific number of characters.

jacobmpeters replied:
I agree. Good suggestion. [f216c85](https://github.com/Analyticsphere/pr2-transformation/pull/1/commits/f216c85e53a845c5e8b7ff0de33ab270bbe4dc28)

eddy-ccc commented on core/utils2.py:
```diff
@@ -0,0 +1,147 @@
+        # Construct query using COUNTIF for efficiency
+        formatted_conditions = ",\n    ".join(
+            f"""
+            (COUNTIF({col} NOT IN ('0', '1') AND {col} IS NOT NULL) = 0) AS `{col}`
```
I'm sure you have this right based on your experience with the source data... but are we really storing these ints as strings??

jacobmpeters replied:
Yes. Unfortunately, the entire database is stored as strings.

eddy-ccc commented on core/utils2.py:
```diff
@@ -0,0 +1,147 @@
+        logger.error(f"Error getting binary fields: {e}")
+        return None, []
+
+def is_false_array(x: str) -> bool:
```
Is str the correct type? The comment below describes it as a list

jacobmpeters replied:
Yes. It's a string. It looks like this `x = '[cid]'`. I should be clear in the comments.

eddy-ccc commented on core/utils2.py:
```diff
@@ -0,0 +1,147 @@
+    """
+    valid_values = [None, "[]", "[178420302]", "[958239616]"]
+
+    # Get unique values; convert to list so order doesn't matter.
```
If order is irrelevant, a set is fine; sets are unordered, but lists are

eddy-ccc commented on core/utils2.py:
```diff
@@ -0,0 +1,147 @@
+    Render a SQL expression to handle survey unwrapping singleton values from variables stored as "false arrays".
+
+    This function returns a SQL snippet that:
+      - Returns NULL when the column's value equals "[]"
```
This comment makes sense - but the variable you're checking against is named col_name. Maybe it should be called col_val instead?

eddy-ccc commented on core/.DS_Store:
Copy the .gitignore from the file process or orchestrator into this project, it'll exclude these .DS_Store files. Preventing them from getting into the git history is 100x easier than removing them after the fact.

jacobmpeters replied:
Done! [8ca35e8](https://github.com/Analyticsphere/pr2-transformation/pull/1/commits/8ca35e8ca0be2a7f1f1a490f51ff25378d6c88c0)

jacobmpeters committed:
add .gitignore

resolves https://github.com/Analyticsphere/pr2-transformation/pull/1#discussion_r1983339686

jacobmpeters committed:
explicitly remove gs:// from path
resolves https://github.com/Analyticsphere/pr2-transformation/pull/1#discussion_r1987378351

jacobmpeters committed:
split output paths
resolves: https://github.com/Analyticsphere/pr2-transformation/pull/1/files/5f56f2575b80cca55af24a6f9e905b42ce02b2aa#r1982170199

jacobmpeters committed:
remove example usage
resolves: https://github.com/Analyticsphere/pr2-transformation/pull/1#discussion_r1982172594

jacobmpeters committed:
move 'allowed extras' to constants module'

jacobmpeters committed:
make seperation of bucket_name, blop_path more clear

jacobmpeters committed:
make seperation of bucket_name, blob_path more clear
